File: src/s0438_find_all_anagrams_in_a_string.rs
/**
 * [438] Find All Anagrams in a String
 *
 * Given two strings s and p, return an array of all the start indices of p's anagrams in s. You may return the answer in any order.
 * 
 * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
 * 
 * 
 * Example:
 *
 *
 * Input:   
 * s = "cbaebabacd"
 * p = "abc"
 *
 * Output: [0,6]
 * 
 * 
 * Input:   
 * s = "abab"
 * p = "ab"
 *
 * Output: [0,1,2]
 */

// problem: https://leetcode.cn/problems/remove-duplicates-from-sorted-array/
use std::collections::HashMap;
pub struct Solution;

impl Solution {
    pub fn find_anagrams(s: String, p: String) -> Vec<i32> {
        let mut needs = HashMap::new();
        p.chars().for_each(|c| { needs.entry(c).and_modify(|e| *e +=1).or_insert(1); } );

        let (mut left, mut right, mut valid) = (0usize, 0usize, 0);
        let mut window = HashMap::new();
        let mut ret = vec![];

        while right < s.len() {
            let r = s.chars().nth(right).unwrap();
            right += 1;
            if needs.contains_key(&r) {
                window.entry(r).and_modify(|e| *e += 1).or_insert(1);
                if window.get(&r) == needs.get(&r) { valid += 1; }
            }
            
            if right - left >= p.len() {
                if valid == needs.len() { ret.push(left as i32); }
                let l = s.chars().nth(left).unwrap();
                left += 1;
                if needs.contains_key(&l) {
                    if window.get(&l) == needs.get(&l) {
                        valid -= 1;
                    }
                    window.entry(l).and_modify(|e| *e -= 1);
                }
            }
        }

        ret
    }
} 

#[test]
fn hashmap_api() {
    let mut map = HashMap::new();
    map.insert('a', 1);
    assert_eq!(map.len(), 1);
    map.entry('a').and_modify(|e| *e += 1 );
    assert_eq!(map.len(), 2);
}

File: cpp/CPP/sliding-window/s0438_find_all_anagrams_in_a_string.cpp
// problem: https://leetcode.cn/problems/find-all-anagrams-in-a-string/
// 示例 1:
// 输入: s = "cbaebabacd", p = "abc"
// 输出: [0,6]
// 解释:
// 起始索引等于 0 的子串是 "cba", 它是 "abc" 的异位词.
// 起始索引等于 6 的子串是 "bac", 它是 "abc" 的异位词.
// 
// 
// 示例 2:
// 输入: s = "abab", p = "ab"
// 输出: [0,1,2]
// 解释:
// 起始索引等于 0 的子串是 "ab", 它是 "ab" 的异位词.
// 起始索引等于 1 的子串是 "ba", 它是 "ab" 的异位词.
// 起始索引等于 2 的子串是 "ab", 它是 "ab" 的异位词.
// 
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;


class Solution {
public:
    vector<int> findAnagrams(string s, string t) {
        vector<int> ret;
        unordered_map<char, int> needs, umap;
        for (auto ch : t) { needs[ch] ++; }
        int sz = s.size();
        int left = 0, right = 0;
        int valid = 0;
        while (right < sz) {
            char rc = s[right];
            right ++;
            if (needs.count(rc)) {
                umap[rc] ++;
                if (umap[rc] == needs[rc]) valid ++;
            }

            while (right - left >= t.size()) {
                if (valid == needs.size()) ret.push_back(left);     // ❗️ 注意
                char lc = s[left];
                left ++;
                if (needs.count(lc)) {
                    if (needs[lc] == umap[lc]) valid --;
                    umap[lc] --;
                }
            }
        }


        return ret;
    }
};


int main() 
{
    Solution s;
    vector<int> ret = s.findAnagrams("baa", "aa");
    for (auto &v : ret) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}